controllers/networkfence_controller.go

```diff
@@ -63,6 +63,11 @@ func validateNetworkFenceSpec(nwFence *csiaddonsv1alpha1.NetworkFence) error {
 	if nwFence.Spec.Cidrs == nil {
 		return errors.New("required parameter cidrs is not specified")
 	}
+
+	if nwFence.Spec.FenceState != csiaddonsv1alpha1.Fenced && nwFence.Spec.FenceState != csiaddonsv1alpha1.Unfenced {
+		return errors.New("invalid fence state. fence state can only be either of Fenced or Unfenced")
+	}
```

Comment:
Will this will be hit as we already have validation at the CR state?

Reply:
maybe if a user had the previous version of the CRD?

Reply:
it breaks the backward compatibility, isn't it? I assume with this change upgrades will also fail as the current spec will not have these fields. @umangachapagain I saw somewhere this recently happened in ocs. can you confirm?

Reply:
Based on the comment [1]  as both NetworkFence CRD and ramen (the consumer of NetworkFence CR as of now) are in v1alpha1 versions, does it require any changes from a compatibility perspective at this point in time?

[1] https://github.com/csi-addons/kubernetes-csi-addons/pull/152#discussion_r864862022

Reply:
add a default state to the CRD if the value is not set so it will be backward compatible

Reply:
I am still not sure how this is expected to work. My knowledge on CRD versioning is very minimal, but I would like to know if the following causes issues:

1. current version of the CRD is deployed, with the current kubernetes-csi-addons controller (without this PR)
2. Ramen gets deployed in a weeks time, expecting the new `FenceState` attribute in the CRD

As the new attribute does not exist in the CRD, Ramen would not be able to access it? How can Ramen require the new version of the CRD/API, or let the user know to update the `NetworkFence` CRD?

Reply:
This is a valid concern. Perhaps it can be incremented to v1alpha2. 

There can be 2 scenarios how this PR can have impacts if the version is not incremented.

1) Installed CRD Version is alpha1 (before this PR) and controller code is new (as of this PR)
    Older specs continue to work i.e. people can submit yamls with older specs. But submitting a yaml that has current changes (fence and unfence states) will not be accepted.

   However, since the current controller refers the fence and unfence state from spec, the controller will report errors related to CR.

2) Installed  CRD is new (as of this PR) and controller is older (before this PR)
    In this scenario, due to the recommendation made in the comment [1], both the older specs  of alpha1 and newer specs should be accepted as no fence state means it will be considered as fenced state by default. Also, from the controller perspective, it does not even look at the spec.State, so it would not report errors related to CR. However, if someone explicitly wants to unfence a cluster by mentioning Unfence state in the spec, the controller would still go ahead and fence them which can be a problem. 

Currently not aware of any consumers who are using the current version of this CR or the controller. So, not sure if this PR would have an impact on any existing deployments or if it is worth the effort to make those changes to increment the version to v1alpha2.

Currently submitting the changes with comments mentioned in [1]. 


[1] https://github.com/csi-addons/kubernetes-csi-addons/pull/152#discussion_r866512971

Reply:
Yes, we dont need to worry about this one now. as we already have a default and the validation is already done in the kube builder. if someone wants new functionality they need to update the controller. keeping that in mind can we remove this check. except for this one all LGTM.

Reply:
ACK. Will be removing this check from the CR validation function.